InventoryIndexer/Model/Queue/RecalculateIndexSalabilityStatus.php

```diff
@@ -0,0 +1,149 @@
+     * @param IndexStructureInterface $indexStructure
+     * @param IndexHandlerInterface $indexHandler
+     * @param AreProductsSalableInterface $areProductsSalable
+     * @param GetStockItemDataInterface $getStockItemData
+     */
+    public function __construct(
+        IndexNameBuilder $indexNameBuilder,
+        DefaultStockProviderInterface $defaultStockProvider,
+        IndexStructureInterface $indexStructure,
+        IndexHandlerInterface $indexHandler,
+        AreProductsSalableInterface $areProductsSalable,
+        GetStockItemDataInterface $getStockItemData
+    ) {
+        $this->indexNameBuilder = $indexNameBuilder;
+        $this->defaultStockProvider = $defaultStockProvider;
+        $this->indexStructure = $indexStructure;
+        $this->indexHandler = $indexHandler;
+        $this->areProductsSalable = $areProductsSalable;
+        $this->getStockItemData = $getStockItemData;
+    }
+
+    /**
+     * @param ReservationData $reservationData
+     *
+     * @return void
+     * @throws StateException
+     */
+    public function execute(ReservationData $reservationData): void
+    {
+        $stockId = $reservationData->getStock();
+        if ($this->defaultStockProvider->getId() === $stockId || !$reservationData->getSkus()) {
+            return;
+        }
+
+        $mainIndexName = $this->indexNameBuilder
+            ->setIndexId(InventoryIndexer::INDEXER_ID)
+            ->addDimension('stock_', (string)$reservationData->getStock())
+            ->setAlias(Alias::ALIAS_MAIN)
+            ->build();
+        if (!$this->indexStructure->isExist($mainIndexName, ResourceConnection::DEFAULT_CONNECTION)) {
+            $this->indexStructure->create($mainIndexName, ResourceConnection::DEFAULT_CONNECTION);
+        }
+        $this->indexHandler->saveIndex(
+            $mainIndexName,
+            $this->getSalabilityData($reservationData->getSkus(), $stockId),
+            ResourceConnection::DEFAULT_CONNECTION
+        );
+    }
+
+    /**
+     * @param string[] $skuList
+     *
+     * @param int $stockId
+     *
+     * @return \Traversable
+     */
+    private function getSalabilityData(array $skuList, int $stockId): \Traversable
+    {
+        $data = array_map(
+            function (IsProductSalableResultInterface $isProductSalableResult) use ($stockId): array {
+                return [
+                    IndexStructure::SKU => $isProductSalableResult->getSku(),
+                    IndexStructure::IS_SALABLE => $isProductSalableResult->isSalable(),
+                    IndexStructure::QUANTITY => $this->getIndexQty($isProductSalableResult->getSku(), $stockId)
+                ];
+            },
+            $this->areProductsSalable->execute($skuList, $stockId)
+        );
+
+        return new \ArrayIterator($data);
+    }
+
+    /**
+     * Get current index QTY value.
+     *
+     * @param string $sku
+     * @param int $stockId
+     *
+     * @return float|null
+     * @throws LocalizedException
```

Comment:
Maybe we catch exception, log it and return `null`?

Reply:
Maybe